File: CombatTracker/CombatTracker.Entities/Current/Actor.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;
using CombatTracker.Entities.Reference;
using CombatTracker.Entities.Reference.Attacks;
using CombatTracker.Base;
using CombatTracker.Entities.Reference.Creatures;
using CombatTracker.Base.Reference;

namespace CombatTracker.Entities.Current
{
    public class Actor
    {
        public Actor()
        {
            //this.Actions = new List<BaseAction>();
            this.Attacks = new List<Attack>();
            this.CriticalEffects = new Stack<CriticalEffect>();
            PercentRequiredAdrenalDB = 0.4;
            PercentAction = 100;
            ConstitutionStat = 90;
        }

        public int ID { get; set; }
        public string Name { get; set; }
        public CharacterType Type { get; set; }
        public string TypeString
        {
            get => Type.ToString();
            set => Type = (CharacterType)Enum.Parse(typeof(CharacterType), value);
        }
        public int Level { get; set; }
        public int HitsTotal { get; set; }
        public int HitsRemaining { get; set; }

        public DamageLevel Damage  {
            get
            {
                var dmg = ((1.0 * HitsRemaining) / (1.0 * HitsTotal))*100;
                if (dmg>=75)
                {
                    return DamageLevel.Healthy;
                }
                if (dmg >= 54)
                {
                    return DamageLevel.Winded;
                }
                if (dmg >= 33)
                {
                    return DamageLevel.Rough;
                }
                if (dmg > 0)
                {
                    return DamageLevel.Bloodied;
                }
                if (dmg > ConstitutionStat)
                {
                    return DamageLevel.Unconscious;
                }
                return DamageLevel.Dead;


            }
        }
        public string DamageString
        {
            get
            {
                return Damage.ToString();
            }
        }
        public int BaseInititive { get; set; }
        public int RolledInititive { get; set; }
        public int ExhaustionTotal { get; set; }
        public int ExhaustionRemaining { get; set; }
        public int PowerPointsTotal { get; set; }
        public int PowerPointsRemaining { get; set; }
        public bool IsConcentrating { get; set; }
        public bool IsMoving { get; set; }
        public bool Suprised { get; set; }
        public bool UsingAdrenalDB { get; set; }
        public double PercentRequiredAdrenalDB { get; set; }
        public double PercentAction { get; set; }
        public int StrengthBonus { get; set; }
        public int CritNegatives { get; set; }
        public double NextSpellTime { get; set; }
        public int ConstitutionStat { get; set; }
        public double Movement { get; set; }
        public string Color { get; set; }


        public int Game_ID { get; set; }
        public int? BaseCreature_ID { get; set; }
        public int? BaseCharacter_ID { get; set; }
        public int? CurrentArmor_ID { get; set; }


        public string CriticalIgnores { get; set; }
        public string CriticalModified { get; set; }




        //public  Character Character { get; set; }
        public  Armor CurrentArmor { get; set; }
        //public  Creature Creature { get; set; }
        //public Game Game { get; set; }

        //public IEnumerable<BaseAction> Actions { get; set; }
        public List<Attack> Attacks { get; set; }
        public Stack<CriticalEffect> CriticalEffects { get; set; }


        public int Inititive => BaseInititive + RolledInititive - 11;

        
        //public IActable Base { get; set; }


        public override string ToString()
        {
            return Name;
        }
        
        //public BaseAction CurrentAction => (from a in Game.GameActions where a.WhoIsActing.ID==ID && a.ActionType == ActionTypeEnum.Current select a).FirstOrDefault();

        //public BaseAction ProposedAction => (from a in Game.GameActions where a.WhoIsActing.ID == ID && a.ActionType == ActionTypeEnum.Proposed select a).FirstOrDefault();

        //public BaseAction FutureAction => (from a in Game.GameActions where a.WhoIsActing.ID == ID && a.ActionType == ActionTypeEnum.Next select a).FirstOrDefault();


        public int StunRounds => (from CR in CriticalEffects where CR.IsStunned select CR).Count();

        public int ParryRounds => (from CR in CriticalEffects where CR.Parry != ParryType.Fine select CR).Count();
        

        public int NegativeRounds => (from CR in CriticalEffects where CR.Negative != 0 select CR).Count();


        public CriticalEffect CurrentCrits => CriticalEffects.LastOrDefault();


        public int HitNegatives {
            get {
                if (HitsRemaining >= 0.75 * HitsTotal)
                {
                    return 0;
                }
                else if (HitsRemaining >= 0.5 * HitsTotal)
                {
                    return -10;
                }
                else if (HitsRemaining >= 0.25 * HitsTotal)
                {
                    return -20;
                }
                else
                {
                    return -30;
                }
            }
        }

        public int ExhNegatives {
            get {
                if (ExhaustionRemaining >= 0.75 * ExhaustionTotal)
                {
                    return 0;
                }
                else if (ExhaustionRemaining >= 0.5 * ExhaustionTotal)
                {
                    return -5;
                }
                else if (ExhaustionRemaining >= 0.25 * ExhaustionTotal)
                {
                    return -15;
                }
                else if (ExhaustionRemaining >= 0.1 * ExhaustionTotal)
                {
                    return -30;
                }
                else if (ExhaustionRemaining >= 0.01 * ExhaustionTotal)
                {
                    return -60;
                }
                else
                {
                    return -100;
                }
            }
        }

        public int Negatives => CritNegatives + HitNegatives + ExhNegatives + (CurrentCrits?.Negative ?? 0);

        public int SpellNegatives
        {
            get
            {
                if (PowerPointsRemaining >= 0.75 * PowerPointsTotal)
                {
                    return 0 + Negatives;
                }
                else if (PowerPointsRemaining >= 0.5 * PowerPointsTotal)
                {
                    return -10 + Negatives;
                }
                else if (PowerPointsRemaining >= 0.25 * PowerPointsTotal)
                {
                    return -20 + Negatives;
                }
                else
                {
                    return -30 + Negatives;
                }
            }
        }




    }
}

File: CombatTracker/CombatTracker.Web/ClientApp/app/entities/current/Actor.ts

module entities.reference {

    // $Classes/Enums/Interfaces(filter)[template][separator]
    // filter (optional): Matches the name or full name of the current item. * = match any, wrap in [] to match attributes or prefix with : to match interfaces or base classes.
    // template: The template to repeat for each matched item
    // separator (optional): A separator template that is placed between all templates e.g. $Properties[public $name: $Type][, ]

    // More info: http://frhagn.github.io/Typewriter/

    
    export class Actor {
        
        // ID
        public id: number = 0;
        // NAME
        public name: string = null;
        // TYPE
        public type: CharacterType = null;
        // TYPESTRING
        public typeString: string = null;
        // LEVEL
        public level: number = 0;
        // HITSTOTAL
        public hitsTotal: number = 0;
        // HITSREMAINING
        public hitsRemaining: number = 0;
        // BASEINITITIVE
        public baseInititive: number = 0;
        // ROLLEDINITITIVE
        public rolledInititive: number = 0;
        // EXHAUSTIONTOTAL
        public exhaustionTotal: number = 0;
        // EXHAUSTIONREMAINING
        public exhaustionRemaining: number = 0;
        // POWERPOINTSTOTAL
        public powerPointsTotal: number = 0;
        // POWERPOINTSREMAINING
        public powerPointsRemaining: number = 0;
        // ISCONCENTRATING
        public isConcentrating: boolean = false;
        // ISMOVING
        public isMoving: boolean = false;
        // SUPRISED
        public suprised: boolean = false;
        // USINGADRENALDB
        public usingAdrenalDB: boolean = false;
        // PERCENTREQUIREDADRENALDB
        public percentRequiredAdrenalDB: number = 0;
        // PERCENTACTION
        public percentAction: number = 0;
        // STRENGTHBONUS
        public strengthBonus: number = 0;
        // CRITNEGATIVES
        public critNegatives: number = 0;
        // NEXTSPELLTIME
        public nextSpellTime: number = 0;
        // CONSTITUTIONSTAT
        public constitutionStat: number = 0;
        // MOVEMENT
        public movement: number = 0;
        // COLOR
        public color: string = null;
        // GAME_ID
        public game_ID: number = 0;
        // BASECREATURE_ID
        public baseCreature_ID: number = null;
        // BASECHARACTER_ID
        public baseCharacter_ID: number = null;
        // CURRENTARMOR_ID
        public currentArmor_ID: number = null;
        // CURRENTARMOR
        public currentArmor: Armor = null;
        // GAME
        public game: Game = null;
        // ATTACKS
        public attacks: Attack[] = [];
        // CRITICALEFFECTS
        public criticalEffects: CriticalEffect[] = [];
        // INITITIVE
        public inititive: number = 0;
        // BASE
        //public base: IActable = null;
        // CURRENTACTION
        public currentAction: BaseAction = null;
        // PROPOSEDACTION
        public proposedAction: BaseAction = null;
        // FUTUREACTION
        public futureAction: BaseAction = null;
        // STUNROUNDS
        public stunRounds: number = 0;
        // PARRYROUNDS
        public parryRounds: number = 0;
        // NEGATIVEROUNDS
        public negativeRounds: number = 0;
        // CURRENTCRITS
        public currentCrits: CriticalEffect = null;
        // NEGATIVES
        public negatives: number = 0;
    }
}